Equipo: 8 
Grupo: 002 
Integrantes: 
1723152, Javier Eduardo Salazar Segura.
1806169, Víctor Hugo Cantú Chávez.
1735532, Damián Atilano Martínez Alvarado.

# 1)	Título de la base de datos.


a)	Nombre con el que se encuentra en la página encontrada.

    International football results from 1872 to 2020
    
b)	Url de la página. 
    https://www.kaggle.com/martj42/international-football-results-from-1872-to-2017
    
    
    
    
    



 # 2) Descripción de los datos.

Tabla con distinta información sobre los resultados de partidos de futbol de selecciones nacionales. 
•	date        object (Cadena de caracteres es object en pandas)
•	local       object
•	visita      object
•	gol_home     int64
•	gol_away     int64
•	torneo      object
•	city        object  
•	pais        object
•	neutral       bool

La mayoría  las columnas tienen  datos que son cadenas de caracteres, ya que son nombres de países, ciudades, torneos y selecciones nacionales. La columna neutral, que se refiere a si el partido se jugo en un país distinto al del equipo local y visitante. Es de valores booleano. Las columnas correspondientes al número de goles ( gol_home, gol_away) son de valores enteros. 

Para el proyecto generamos 3 bases de datos a partir de esta. Pero todas conservan el mismo formato, nombre de columnas y tipos de datos por columna




Se revisó que no haya datos nulos que entorpezcan la base de datos usando: 
 
Data.isnull.any(); 

date        False   
local       False
visita      False
gol_home    False
gol_away    False
torneo      False
city        False
pais        False
neutral     False
dtype: bool

No hay datos nulos en la base de datos principal ni ninguna de las 3 bases de datos que generamos a partir de esta.

Se decidió no eliminar ninguna columna ya que todas tienen datos relevantes y ninguna tiene algún inconveniente que entorpezca el estudio como datos nulos o filas vacías. Por eso se conservarán todas las columnas.
 
 



# 3)	Justificación del uso de datos. 

Es una base de datos muy completa para lo que queremos analizar tiene todos los datos que podemos analizar fácilmente sobre un partido de futbol, desde el numero de goles, la fecha, el torneo y el lugar exacto donde se jugó. 
Es una base de datos muy bien ordenada y consistente con información correcta para desarrollar proyectos de minería de datos que busquen patrones relacionados al resultado de un partido de futbol. 

Además de ser información muy útil y completa, es de dominio público y tiene una valoración de 10 en su usabilidad en la página de Kaggle, es una base de datos apta para obtener la información que pretendemos usar.



# 4)	Objetivo final. 

El objetivo principal es encontrar patrones que históricamente influyan en el resultado de un partido de futbol de la selección mexicana. Para poder así realizar pronósticos del resultado de un partido de México conociendo las condiciones como: País donde se juega, partido amistoso, torneo, ser local visitante o neutral, etc. 

El objetivo secundario sería poder encontrar patrones para pronosticar el resultado de un partido de México contra ciertos rivales de interés ya sea por su nivel internacional o por rivalidad que puedan despertar un interés especial, como por ejemplo México vs Estados Unidos, México vs Argentina y México vs Brasil. Esto usando los datos históricos de México jugando contra este tipo de países.



# 5)	Preguntas de interés

1- ¿Qué tan cierto es que la selección mexicana no suele perder jugando como local? Revisando el caso donde se juega en cualquier lugar de México y el caso espcifico .

2- ¿Qué tan frecuente es que México gane un partido contra Estados Unidos, Argentina o Brasil? Según los datos históricos.

En general se busca pronosticar el resultado de un partido de México como local. Y los casos especificos contra las selecciones de Estados Unidos, Argentina y Brasil. Esto respondería las dos preguntas de interés.



# 6)    Parte de progrmación 

   Primero importamos las librerías pandas y numpy.
   Además se hace la lectura del archivo CSV que se descargó desde Kaggle.
   Se usa datafr.head() para verificar la lectura correcta.

In [1]:
import pandas as pd
import numpy as np
datafr=pd.read_csv("results.csv")
datafr.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

 Se procede a cambiar los nombres de algunas columnas para hacerlos más cortos y fácil de entender. Guardando el nuevo datafrme con los nombres nuevos en las columnas en la variable datasp.
 Se revisa que no haya datos nulos en los datos originales.

In [3]:
# Haciendo los nombres de las columnas más cortos.
datasp=datafr.rename(columns = {'home_team':'local','away_team':'visita','home_score':'gol_home','away_score':'gol_away','tournament':'torneo','tournament':'torneo','country':'pais'})
#### buscando datos nulos para eliminarlos ya que afectan el analisis
datasp.isnull().any()




date        False
local       False
visita      False
gol_home    False
gol_away    False
torneo      False
city        False
pais        False
neutral     False
dtype: bool

Confirmado que no hay datos nulos, ahora se verificará que el tipo de dato tenga sentido con lo registrado en la tabla.

In [71]:
# No hay datos nulos
# viendo tipos de datos de las columnas
datasp.dtypes

date        object
local       object
visita      object
gol_home     int64
gol_away     int64
torneo      object
city        object
pais        object
neutral       bool
dtype: object

Se confirma que todos los tipos de datos son coherentes con lo registrado en la tabla. Se compone de cadenas de caracteres (object), de valores enteros (int64) para el número de goles de cada equipo.
Además se muestra el contenido de la variable datasp para verificar el cambio de nombre en las columnas.

In [72]:
## todos los tipos de datos son correctos
### observando cambio de nombre en las columnas
datasp.head()

date     local    visita  gol_home  gol_away    torneo     city  \
0  1872-11-30  Scotland   England         0         0  Friendly  Glasgow   
1  1873-03-08   England  Scotland         4         2  Friendly   London   
2  1874-03-07  Scotland   England         2         1  Friendly  Glasgow   
3  1875-03-06   England  Scotland         2         2  Friendly   London   
4  1876-03-04  Scotland   England         3         0  Friendly  Glasgow   

       pais  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

Se filtran los datos originales para obtener unicamente los datos de los partidos de la selección de México, estos datos se usarán para el estudio que busca responder la pregunta de interés 1.

Se destaca el uso de la función pd.concat. Esta solución se encontró en el foro de stackoverflow. 
Se anexa la url: https://stackoverflow.com/questions/12850345/how-do-i-combine-two-dataframes

In [10]:
## filtrando para obtener solo los partidos de la seleccion de MEXICO ya que es el equipo del cual analizaremos sus resultados
datasp[datasp['local']=='Mexico']
datasp[datasp['visita']=='Mexico']
datamx=pd.concat([datasp[datasp['local']=='Mexico'],datasp[datasp['visita']=='Mexico'] ], ignore_index=True, sort=False)
datamx.head()
datamx.to_csv('Solo_Mex.csv', index=False)
# Se encontro la forma para concatenar los data frame en https://stackoverflow.com/questions/12850345/how-do-i-combine-two-dataframes

Se usa data.shape para verificar la reducción de la base datos respecto a la original.

In [75]:
# viendo el numero de ilas y columnas de la base de datos de Mexico
datamx.shape


(873, 9)

A continuación se muestra el contenido de datamx para verificar que estén los datos de los partidos de México tanto de local como de visitante.

También se verifica que no haya datos nulos en la nueva tabla, se compara las dimensiones de datasp y datamx. Se verifican los tipos de datos de datamx.

In [77]:
#Observando la base de datos de mexico
datamx.head()

date   local     visita  gol_home  gol_away  \
0  1923-12-09  Mexico  Guatemala         2         1   
1  1923-12-12  Mexico  Guatemala         2         0   
2  1923-12-16  Mexico  Guatemala         3         3   
3  1934-03-04  Mexico       Cuba         3         2   
4  1934-03-11  Mexico       Cuba         5         0   

                         torneo         city    pais  neutral  
0                      Friendly  Mexico City  Mexico    False  
1                      Friendly  Mexico City  Mexico    False  
2                      Friendly  Mexico City  Mexico    False  
3  FIFA World Cup qualification  Mexico City  Mexico    False  
4  FIFA World Cup qualification  Mexico City  Mexico    False

In [78]:
#Observando la base de datos de mexico
datamx.tail()

date        local  visita  gol_home  gol_away  \
868  2019-09-10    Argentina  Mexico         4         0   
869  2019-10-11      Bermuda  Mexico         1         5   
870  2019-11-15       Panama  Mexico         0         3   
871  2020-10-07  Netherlands  Mexico         0         1   
872  2020-11-17        Japan  Mexico         0         2   

                      torneo         city           pais  neutral  
868                 Friendly  San Antonio  United States     True  
869  CONCACAF Nations League     Hamilton        Bermuda    False  
870  CONCACAF Nations League  Panama City         Panama    False  
871                 Friendly    Amsterdam    Netherlands    False  
872                 Friendly         Graz        Austria     True

In [79]:
# La base de datos datamx si incluye los datos de mexico tanto de local como de visita.
# se revisara que no haya nulos y que el tipo de datos se haya mantenido
# buscando valores nulos
datamx.isnull().any()

date        False
local       False
visita      False
gol_home    False
gol_away    False
torneo      False
city        False
pais        False
neutral     False
dtype: bool

In [80]:
#revisando tipo de datos
datamx.dtypes
# observamos que no se tiene datos nulos y que los tipos de datos se mantuvieron por lo que los datos se filrtraron correctamente

date        object
local       object
visita      object
gol_home     int64
gol_away     int64
torneo      object
city        object
pais        object
neutral       bool
dtype: object

In [81]:
datasp.count

<bound method DataFrame.count of              date          local       visita  gol_home  gol_away  \
0      1872-11-30       Scotland      England         0         0   
1      1873-03-08        England     Scotland         4         2   
2      1874-03-07       Scotland      England         2         1   
3      1875-03-06        England     Scotland         2         2   
4      1876-03-04       Scotland      England         3         0   
...           ...            ...          ...       ...       ...   
41871  2020-11-18         Greece     Slovenia         0         0   
41872  2020-11-18        Albania      Belarus         3         2   
41873  2020-11-18     Kazakhstan    Lithuania         1         2   
41874  2020-12-04          Qatar   Bangladesh         5         0   
41875  2020-12-09  United States  El Salvador         6         0   

                             torneo             city           pais  neutral  
0                          Friendly          Glasgow       

In [82]:
datamx.count

<bound method DataFrame.count of            date        local     visita  gol_home  gol_away  \
0    1923-12-09       Mexico  Guatemala         2         1   
1    1923-12-12       Mexico  Guatemala         2         0   
2    1923-12-16       Mexico  Guatemala         3         3   
3    1934-03-04       Mexico       Cuba         3         2   
4    1934-03-11       Mexico       Cuba         5         0   
..          ...          ...        ...       ...       ...   
868  2019-09-10    Argentina     Mexico         4         0   
869  2019-10-11      Bermuda     Mexico         1         5   
870  2019-11-15       Panama     Mexico         0         3   
871  2020-10-07  Netherlands     Mexico         0         1   
872  2020-11-17        Japan     Mexico         0         2   

                           torneo         city           pais  neutral  
0                        Friendly  Mexico City         Mexico    False  
1                        Friendly  Mexico City         Mexico   

Se observa que en efecto, se redujo la cantidad de datos, se tienen 873 partidos de la selección de México en la base de datos de 41876 partidos. Además no hubo datos nulos después de generar datamx y los tipos de datos continuan siendo los mismos.

In [1]:
# observamos la reduccion tras el filtrado de los datos. Ahora podemos usar data mx para trabajar con datos de la seleccion de mexico
# tambien agruparemos los partidos de interes: Mexico vs USA, Mexico vs Brasil, Mexico vs Argentina

Ahora se agrupará y generaraá los archivos que tengan solamente los partidos de Mexico vs USA, Mexico vs Brasil y Mexico vs Argentina. Estos archivos serán de utilidad para el estudio que busca resolver la pregunta de interés 2.

In [11]:
datamxus=pd.concat([datamx[datamx['local']=='United States'],datamx[datamx['visita']=='United States'] ], ignore_index=True, sort=False)

datamxus.head()

date          local  visita  gol_home  gol_away  \
0  1957-04-28  United States  Mexico         2         7   
1  1960-11-06  United States  Mexico         3         3   
2  1965-03-07  United States  Mexico         2         2   
3  1972-09-10  United States  Mexico         1         2   
4  1974-09-08  United States  Mexico         0         1   

                         torneo         city           pais  neutral  
0  FIFA World Cup qualification   Long Beach  United States    False  
1  FIFA World Cup qualification  Los Angeles  United States    False  
2  FIFA World Cup qualification  Los Angeles  United States    False  
3  FIFA World Cup qualification  Los Angeles  United States    False  
4                      Friendly       Dallas  United States    False

In [12]:
datamxus.tail()

date   local         visita  gol_home  gol_away  \
61  2005-03-27  Mexico  United States         2         1   
62  2009-08-12  Mexico  United States         2         1   
63  2012-08-15  Mexico  United States         0         1   
64  2013-03-26  Mexico  United States         0         0   
65  2017-06-11  Mexico  United States         1         1   

                          torneo         city    pais  neutral  
61  FIFA World Cup qualification  Mexico City  Mexico    False  
62  FIFA World Cup qualification  Mexico City  Mexico    False  
63                      Friendly  Mexico City  Mexico    False  
64  FIFA World Cup qualification  Mexico City  Mexico    False  
65  FIFA World Cup qualification  Mexico City  Mexico    False

In [13]:
datamxar=pd.concat([datamx[datamx['local']=='Argentina'],datamx[datamx['visita']=='Argentina'] ], ignore_index=True, sort=False)
datamxar.head()

date      local  visita  gol_home  gol_away  \
0  1930-07-19  Argentina  Mexico         6         3   
1  1960-03-10  Argentina  Mexico         3         2   
2  1960-03-17  Argentina  Mexico         2         0   
3  1962-03-28  Argentina  Mexico         1         0   
4  1984-10-25  Argentina  Mexico         1         1   

                      torneo          city        pais  neutral  
0             FIFA World Cup    Montevideo     Uruguay     True  
1  Pan American Championship      San José  Costa Rica     True  
2  Pan American Championship      San José  Costa Rica     True  
3                   Friendly  Buenos Aires   Argentina    False  
4                   Friendly  Buenos Aires   Argentina    False

In [14]:
datamxar.tail()

date   local     visita  gol_home  gol_away              torneo  \
26  2005-03-09  Mexico  Argentina         1         1            Friendly   
27  2005-06-26  Mexico  Argentina         1         1  Confederations Cup   
28  2007-07-11  Mexico  Argentina         0         3        Copa América   
29  2008-06-04  Mexico  Argentina         1         4            Friendly   
30  2015-09-08  Mexico  Argentina         2         2            Friendly   

            city           pais  neutral  
26   Los Angeles  United States     True  
27       Hanover        Germany     True  
28  Puerto Ordaz      Venezuela     True  
29     San Diego  United States     True  
30     Arlington  United States     True

In [15]:
datamxbr=pd.concat([datamx[datamx['local']=='Brazil'],datamx[datamx['visita']=='Brazil'] ], ignore_index=True, sort=False)
datamxbr.head()

date   local  visita  gol_home  gol_away                     torneo  \
0  1950-06-24  Brazil  Mexico         4         0             FIFA World Cup   
1  1952-04-06  Brazil  Mexico         2         0  Pan American Championship   
2  1954-06-16  Brazil  Mexico         5         0             FIFA World Cup   
3  1960-03-06  Brazil  Mexico         2         2  Pan American Championship   
4  1960-03-15  Brazil  Mexico         2         1  Pan American Championship   

             city         pais  neutral  
0  Rio de Janeiro       Brazil    False  
1        Santiago        Chile     True  
2          Geneva  Switzerland     True  
3        San José   Costa Rica     True  
4        San José   Costa Rica     True

In [16]:
datamxbr.tail()

date   local  visita  gol_home  gol_away              torneo  \
36  2004-07-18  Mexico  Brazil         0         4        Copa América   
37  2005-06-19  Mexico  Brazil         1         0  Confederations Cup   
38  2007-09-12  Mexico  Brazil         1         3            Friendly   
39  2011-10-11  Mexico  Brazil         1         2            Friendly   
40  2012-06-03  Mexico  Brazil         2         0            Friendly   

          city           pais  neutral  
36       Piura           Peru     True  
37     Hanover        Germany     True  
38  Foxborough  United States     True  
39     Torreón         Mexico    False  
40   Arlington  United States     True

In [17]:
# Guardando archivos condatos filtrados
datamxus.to_csv('Solo_MexUS.csv', index=False)
datamxar.to_csv('Solo_MexAR.csv', index=False)
datamxbr.to_csv('Solo_MexBR.csv', index=False)

Se generaron correctamente los archivos necesarios. Se verificó que las variables datamxus,datamxar y datamxbr tengan los partidos correspondientes tanto de local como de visita. 